File: sdks/typescript/packages/core/src/token-usage.ts
import { TokenUsage } from "./events";

/**
 * Map a LangChain-family `usage_metadata` object into an AG-UI {@link TokenUsage}.
 *
 * LangChain and LangGraph both attach usage as `{ input_tokens, output_tokens,
 * total_tokens, input_token_details: { cache_read }, output_token_details:
 * { reasoning } }`. This maps only those numeric counts plus optional
 * provider/model labels — never prompt/completion content. Returns `undefined`
 * when no usage metadata is present, so callers can omit usage rather than
 * report zeros.
 */
export function tokenUsageFromLangChainMetadata(
  usageMetadata: any,
  { provider, model }: { provider?: string; model?: string },
): TokenUsage | undefined {
  if (!usageMetadata) return undefined;

  const inputDetails = usageMetadata.input_token_details ?? {};
  const outputDetails = usageMetadata.output_token_details ?? {};

  const entry: TokenUsage = {};
  if (provider != null) entry.provider = provider;
  if (model != null) entry.model = model;
  if (usageMetadata.input_tokens != null) entry.inputTokens = usageMetadata.input_tokens;
  if (usageMetadata.output_tokens != null) entry.outputTokens = usageMetadata.output_tokens;
  if (usageMetadata.total_tokens != null) entry.totalTokens = usageMetadata.total_tokens;
  if (outputDetails.reasoning != null) entry.reasoningTokens = outputDetails.reasoning;
  if (inputDetails.cache_read != null) entry.cachedInputTokens = inputDetails.cache_read;

  return entry;
}

/**
 * Map an AI-SDK (v5) `LanguageModelUsage` object into an AG-UI {@link TokenUsage}.
 *
 * AI-SDK's keys already match: `inputTokens`, `outputTokens`, `totalTokens`,
 * `reasoningTokens`, `cachedInputTokens`. AI-SDK reports `NaN`/`undefined` for
 * counts a provider didn't return, so only finite numbers are copied. Returns
 * `undefined` when no finite count is present (so callers omit empty usage).
 */
export function tokenUsageFromAiSdkUsage(
  usage: any,
  { provider, model }: { provider?: string; model?: string },
): TokenUsage | undefined {
  if (!usage) return undefined;

  const num = (v: unknown): number | undefined =>
    typeof v === "number" && Number.isFinite(v) ? v : undefined;

  const counts: Partial<Record<keyof TokenUsage, number | undefined>> = {
    inputTokens: num(usage.inputTokens),
    outputTokens: num(usage.outputTokens),
    totalTokens: num(usage.totalTokens),
    reasoningTokens: num(usage.reasoningTokens),
    cachedInputTokens: num(usage.cachedInputTokens),
  };
  if (!Object.values(counts).some((v) => v !== undefined)) return undefined;

  const entry: TokenUsage = {};
  if (provider != null) entry.provider = provider;
  if (model != null) entry.model = model;
  for (const [key, value] of Object.entries(counts)) {
    if (value !== undefined) (entry as any)[key] = value;
  }
  return entry;
}

const COUNT_FIELDS = [
  "inputTokens",
  "outputTokens",
  "totalTokens",
  "reasoningTokens",
  "cachedInputTokens",
] as const;

/**
 * Sum per-call {@link TokenUsage} entries into one entry per `(provider, model)`
 * pair. Order follows first appearance. A count field stays `undefined` when no
 * member of the group reported it, so "not reported" stays distinct from zero.
 *
 * Protocol-agnostic: works on any producer's `TokenUsage[]`, so integrations
 * share it rather than reimplementing aggregation.
 */
export function aggregateTokenUsage(entries: TokenUsage[]): TokenUsage[] {
  const grouped = new Map<string, TokenUsage>();

  for (const entry of entries) {
    const key = `${entry.provider ?? ""} ${entry.model ?? ""}`;
    let target = grouped.get(key);
    if (!target) {
      target = { provider: entry.provider, model: entry.model };
      grouped.set(key, target);
    }
    for (const field of COUNT_FIELDS) {
      const value = entry[field];
      if (value == null) continue;
      target[field] = (target[field] ?? 0) + value;
    }
  }

  return [...grouped.values()];
}

File: sdks/community/rust/crates/ag-ui/src/token_usage.rs
//! Per-model token accounting carried by `RUN_FINISHED` and `RUN_ERROR`.

use serde::{Deserialize, Serialize};

/// A numeric-only token usage summary for one `(provider, model)` pair.
///
/// Deliberately carries nothing identifying or content-bearing — no prompts,
/// completions, message text, or thread/run/user ids. Only provider and model
/// labels plus counts, so usage can be logged and aggregated in places where
/// conversation content must not go.
///
/// Every count is optional and every count is a non-negative integer: `None`
/// means *the provider did not report it*, which is distinct from zero.
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
#[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
pub struct TokenUsage {
    /// The inference provider, for example `"anthropic"`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    /// The model id, for example `"claude-opus-5"`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    /// Tokens consumed by the prompt.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub input_tokens: Option<u64>,
    /// Tokens produced by the model.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output_tokens: Option<u64>,
    /// Total tokens, as reported by the provider (not necessarily the sum of
    /// the other fields).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub total_tokens: Option<u64>,
    /// Output tokens spent on reasoning.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reasoning_tokens: Option<u64>,
    /// Input tokens served from the provider's prompt cache.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cached_input_tokens: Option<u64>,
}

impl TokenUsage {
    /// An empty usage entry.
    pub fn new() -> Self {
        Self::default()
    }

    /// Whether any count is present. A labels-only entry reports nothing and
    /// should be omitted rather than emitted.
    pub const fn has_counts(&self) -> bool {
        self.input_tokens.is_some()
            || self.output_tokens.is_some()
            || self.total_tokens.is_some()
            || self.reasoning_tokens.is_some()
            || self.cached_input_tokens.is_some()
    }
}

/// Sums per-call [`TokenUsage`] entries into one entry per `(provider, model)`
/// pair, in order of first appearance.
///
/// A count stays `None` when no member of a group reported it, keeping "not
/// reported" distinct from zero.
///
/// ```
/// # use ag_ui::{TokenUsage, aggregate_token_usage};
/// let calls = vec![
///     TokenUsage { model: Some("m".into()), input_tokens: Some(10), ..Default::default() },
///     TokenUsage { model: Some("m".into()), input_tokens: Some(5), ..Default::default() },
/// ];
/// let total = aggregate_token_usage(&calls);
/// assert_eq!(total.len(), 1);
/// assert_eq!(total[0].input_tokens, Some(15));
/// assert_eq!(total[0].output_tokens, None);
/// ```
pub fn aggregate_token_usage(entries: &[TokenUsage]) -> Vec<TokenUsage> {
    let mut grouped: Vec<TokenUsage> = Vec::new();

    for entry in entries {
        let index = match grouped
            .iter()
            .position(|g| g.provider == entry.provider && g.model == entry.model)
        {
            Some(index) => index,
            None => {
                grouped.push(TokenUsage {
                    provider: entry.provider.clone(),
                    model: entry.model.clone(),
                    ..Default::default()
                });
                grouped.len() - 1
            }
        };

        let target = &mut grouped[index];
        add_into(&mut target.input_tokens, entry.input_tokens);
        add_into(&mut target.output_tokens, entry.output_tokens);
        add_into(&mut target.total_tokens, entry.total_tokens);
        add_into(&mut target.reasoning_tokens, entry.reasoning_tokens);
        add_into(&mut target.cached_input_tokens, entry.cached_input_tokens);
    }

    grouped
}

/// Adds `value` into `target`, leaving `target` untouched when nothing was
/// reported. Saturating, so a hostile producer cannot panic the aggregation.
fn add_into(target: &mut Option<u64>, value: Option<u64>) {
    if let Some(value) = value {
        *target = Some(target.unwrap_or(0).saturating_add(value));
    }
}
